IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c

```
@@ -1,51 +1,447 @@
+  BOOLEAN                         bResultFlag = FALSE;
+  IPMI_SELF_TEST_RESULT_RESPONSE  *SelfTestResult;
+
+  SelfTestResult = (VOID *)&IpmiInstance->TempData[0];
+
+  //
+  // Get the SELF TEST Results.
+  //
+  //
+  // Note: If BMC PcdIpmiBmcReadyDelayTimer < BMC_IPMI_TIMEOUT, it need set Retries as 1. Otherwise it will make SELT failure, caused by below condition (EFI_ERROR(Status) || Retries == 0)
+  //
+  if (PcdGet8 (PcdIpmiBmcReadyDelayTimer) < PcdGet8 (PcdBmcTimeoutSeconds)) {
+    Retries = 1;
+  } else {
+    Retries = PcdGet8 (PcdIpmiBmcReadyDelayTimer);
+  }
+
+  DataSize = sizeof (IpmiInstance->TempData);
+
+  IpmiInstance->TempData[1] = 0;
+
+  do {
+    Status = IpmiSendCommand (
+               &IpmiInstance->IpmiTransport,
+               IPMI_NETFN_APP,
+               0,
+               IPMI_APP_GET_SELFTEST_RESULTS,
+               NULL,
+               0,
+               IpmiInstance->TempData,
+               &DataSize
+               );
+
+    if (Status == EFI_SUCCESS) {
+      switch (SelfTestResult->Result) {
+        case IPMI_APP_SELFTEST_NO_ERROR:
+        case IPMI_APP_SELFTEST_NOT_IMPLEMENTED:
+        case IPMI_APP_SELFTEST_ERROR:
+        case IPMI_APP_SELFTEST_FATAL_HW_ERROR:
+          bResultFlag = TRUE;
+          break;
+
+        default:
+          break;
+      } // switch
+
+      if (bResultFlag) {
+        break;
+      }
+    }
+
+    MicroSecondDelay (500 * 1000);
+  } while (--Retries > 0);
+
+  //
+  // If Status indicates a Device error, then the BMC is not responding, so send an error.
+  //
+  if (EFI_ERROR (Status) || (Retries == 0)) {
+    DEBUG ((DEBUG_ERROR, "\n[IPMI]  BMC self-test does not respond (status: %r)!\n\n", Status));
+    if (*ErrorCount < MAX_SOFT_COUNT) {
+      StatusCodeValue[*ErrorCount] = EFI_COMPUTING_UNIT_FIRMWARE_PROCESSOR | EFI_CU_FP_EC_COMM_ERROR;
+      (*ErrorCount)++;
+    }
+
+    IpmiInstance->BmcStatus = BMC_HARDFAIL;
+    return Status;

```

Probably want to set it to timeout error if the retry count expires?

---

I actually think its better to bubble up the underlying error code from then an error indicating we ran out of retries personally.